### DISCUSSION TITLE
Many small enhancement request

### DISCUSSION THREAD

**Discussion by zenzeinet** [TabularEditor/TabularEditor3#1103] [Ideas]
Hi,
After finally upgrading to the Enterprise version, we've collected a number of improvement items, most of them short. Hope they make it in future versions! Here's the ordered list:

TOM Explorer: 
- Add [expand all] button
- Add [expand all] and [collapse all] to menu and to right-click and shortcuts
- Create relationships from column right click >> missing tables/columns like in TE2 to select from/to related column
- ~When columns dont fit: could we have scroll bar instead of all-compressed? (like TE2)~ [available in 3.3.3](https://github.com/TabularEditor/TabularEditor3/releases/tag/3.3.3)

Windows Manager:
- ~~Remove animation from pinned/unpinned windows (way too slow animation, at least on our systems), allow it to be instantaneous.~~ [available in 3.3.4](https://docs.tabulareditor.com/te3/other/release-notes/3_3_4.html)

Shortcuts:
- ~~Dedicated shortcut / menu item for Save Model (vs save file / dax query...). Ctrl-S (or something else) should be exclusively dedicated to this.~~ [available in 3.3.4](https://docs.tabulareditor.com/te3/other/release-notes/3_3_4.html)
- ~~Dedicated shortcut and menu item for [format dax (debug commas)]~~ [available in 3.3.4](https://docs.tabulareditor.com/te3/other/release-notes/3_3_4.html)
- Shortcut for collapseAll , expandAll (to avoid swapping to mouse) like in TE2

Dax query: 
- when result is short, if no ORDER BY is put in the DAX query, then automatically sort by default the results by (column1, column2, column3)
- when a grid result is manually ordered (via click), keep it ordered: right now ordering is lost if you re-execute the query
- Don't force to close DAX Queries when changing models. 
        - Make it really painful to change from dev to prod models, or in general any change in connection.

Dependencies Window:
- Would be better to have a tree (rather than radial buttons) with 
	- Object that depend > collapsable
	- Object on which this depends > collapsable
	- (new) relationships to/from > collapsable (multi-level) 

Diagram view 
- Already posted some improvements that could be done in github
	- In general, take inspiration from PBID which covers many of these points

Functionality:
- "Show History" of TE2 missing?
- ~Options: is there an option like the "Automatic Formula Fix-up" in TE2? This is needed to disable fix-up in the (few, but sometimes extremely useful) cases when fix-up is not wanted.~ [available in 3.3.3](https://github.com/TabularEditor/TabularEditor3/releases/tag/3.3.3)

Expression Editor, editing measures:
- The Format String property is very often changed. Add it to the Expression editor bar, to easily change it without having the properties window always open

Hope you can implement these. If some of them are already covered, please tell us how!

**Comment by otykier** [comment]
Many great suggestions as always (going forward, please create separate issues for each, to make it easier to track delivery and provide comments/clarifying questions for each).

> Shortcut for collapseAll , expandAll (to avoid swapping to mouse) like in TE2

This is already possible: CTRL+Right arrow to expand all (from current node), and CTRL+Left arrow to collapse all. [Full shortcuts](https://docs.tabulareditor.com/te3/features/shortcuts.html#tom-explorer).

> when result is short, if no ORDER BY is put in the DAX query, then automatically sort by default the results by (column1, column2, column3)

This should be an option, in my opinion, as users might be confused why a similar query in SSMS or DAX Studio returns a result that is ordered differently.

> The Format String property is very often changed. Add it to the Expression editor bar, to easily change it without having the properties window always open

I'm hesitant to use the Expression Editor for anything but DAX, SQL and M. Otherwise, I'm not sure where to draw the limit - should we also allow editing Descriptions, Display Folders, and other string properties this way? For calculated objects, you can use DAX Scripts, as these include a few more properties than just the DAX expression, namely format strings, descriptions and display folders.

**Comment by otykier** [comment]
A few of these have been fixed in [3.3.3](https://github.com/TabularEditor/TabularEditor3/releases/tag/3.3.3) (I striked out the ones in your post). @zenzeinet are you happy with the shortcuts for collapse all / expand all as described above?

**Comment by otykier** [comment]
Some more fixes are available in [3.3.4](https://docs.tabulareditor.com/te3/other/release-notes/3_3_4.html).

**Comment by zenzeinet** [comment]
Hi Daniel,
Thanks for the fixes! - hope more get implemented. A few things:

- You asked: Collapse / Expand: Good, but there is still one small inconvenient in keyboard: when the TOM Explorer has a side-to-side scrollbar (i.e. is too narrow), the "Right" arrow key expands... but also scrolls to the right!, making the objects not visible

- I have a long list of potential improvements... let's open many :)

**Comment by otykier** [comment]
> I have a long list of potential improvements... let's open many :)

Thanks a ton for these! And please do keep them coming!
